lib/bakeAmbientOcclusion.js

```diff
@@ -11,53 +11,495 @@ var Matrix3 = Cesium.Matrix3;
+        buffer: aoBufferId,
+        byteOffset: 0,
+        byteLength: allAODataLength * 4,
+        target: WebGLConstants.ARRAY_BUFFER
+    };
+
+    // add accessor for each primitive
+    var primitiveCount = primitiveOrder.length;
+    var byteOffset = 0;
+    for (var i = 0; i < primitiveCount; i++) {
+        var primitive = primitiveOrder[i];
+        var primitiveVertexCount = parameters.primitiveVertexCounts[i];
+        var name = 'accessor_' + parameters.primitiveNames[i] + '_AO';
+        primitive.attributes._OCCLUSION = name;
+        gltf.accessors[name] = {
+            bufferView: aoBufferViewId,
+            byteOffset: byteOffset,
+            componentType: WebGLConstants.FLOAT,
+            count: primitiveVertexCount,
+            type: 'SCALAR'
+        };
+        byteOffset += primitiveVertexCount * 4;
+    }
+}
+
+function addAoToShaders(options) {
+    var gltf = options.gltf;
+    var techniqueIds = options.techniqueIds;
+    var shaderMode = options.shaderMode;
+    var ambientShadowContribution = options.ambientShadowContribution;
+
+    // Keep track of which shaders have been edited
+    var shaders = gltf.shaders;
+    var shadersEdited = {};
+    for (var shaderID in shaders) {
+        if (shaders.hasOwnProperty(shaderID)) {
+            shadersEdited[shaderID] = false;
+        }
+    }
+
+    // For each technique,
+    var techniquesCount = techniqueIds.length;
+    for (var i = 0; i < techniquesCount; i++) {
+        var technique = gltf.techniques[techniqueIds[i]];
+
+        var glslNewAttributes = 'attribute float a_ambientOcclusion; \n'; // snippet for adding attributes
+        var glslNewVaryings = 'varying float v_ambientOcclusion; \n'; // snippet for adding varying
+        var glslPassThrough = 'v_ambientOcclusion = a_ambientOcclusion; \n'; // snippet for passing values from vs to fs
+        var glslChangeColor; // snippet for editing the final glsl color
+        var glslDiffuseColor; // snippet for accessing the diffuse color
+
+        if (shaderMode !== 'multiply') {
+            // Check for a diffuse uniform and note if it's a texture or a term
+            var diffuseParameter = technique.parameters.diffuse;
+            if (!defined(diffuseParameter)) {
+                throw new DeveloperError('Could not find parameter diffuse in technique ' + techniqueIds[i]);
+            }
+
+            var uniforms = technique.uniforms;
+            // fetch the uniform name for the diffuse uniform
+            for (var uniformName in uniforms) {
+                if (uniforms.hasOwnProperty(uniformName)) {
+                    if (uniforms[uniformName] === 'diffuse') {
+                        glslDiffuseColor = uniformName;
+                        break;
+                    }
+                }
+            }
+            if (diffuseParameter.type === WebGLConstants.SAMPLER_2D) {
+                glslNewVaryings += 'varying vec2 v_ambientOcclusionUV; \n';
```

Comment:
As we talked about offline, it would be better to reuse the shader's texcoord varying if we can get its name.